src/rust/engine/process_executor/src/test_server.rs

```diff
@@ -0,0 +1,159 @@
+  ///
+  ///                      If an ExecuteRequest request is received which is not equal to this
+  ///                      MockExecution's execute_request, an error will be returned.
+  ///
+  ///                      If a GetOperation request is received whose name is not equal to this
+  ///                      MockExecution's name, or more requests are received than stub responses
+  ///                      are available for, an error will be returned.
+  pub fn new(mock_execution: MockExecution) -> TestServer {
+    let mock_server = MockServer::new(mock_execution);
+
+    let env = Arc::new(grpcio::Environment::new(1));
+    let mut server_transport = grpcio::ServerBuilder::new(env)
+        .register_service(bazel_protos::remote_execution_grpc::create_execution(mock_server.clone()))
+        .register_service(bazel_protos::operations_grpc::create_operations(mock_server.clone()))
+        .bind("localhost", 0)
+        .build()
+        .unwrap();
+    server_transport.start();
+
+    TestServer {
+      mock_server,
+      server_transport,
+    }
+  }
+
+  ///
+  /// The address on which this server is listening over insecure HTTP transport.
+  ///
+  pub fn address(&self) -> String {
+    let bind_addr = self.server_transport.bind_addrs().first().unwrap();
+    format!("{}:{}", bind_addr.0, bind_addr.1)
+  }
+}
+
+impl Drop for TestServer {
+  fn drop(&mut self) {
+    let expected_responses = self.mock_server.mock_execution.operation_responses.len();
+    let issued_responses = self.mock_server.next_operation_response.lock().unwrap().deref().clone();
+    assert_eq!(
+      expected_responses,
+      issued_responses,
+      "Expected {} responses but only issued {}. Expected responses:\n{}\nReceived requests:\n{}",
+      expected_responses,
+      issued_responses,
+      MockServer::display_all(self.mock_server.mock_execution.operation_responses.iter()),
+      MockServer::display_all(self.mock_server.received_messages.deref().lock().unwrap().iter())
+    )
+  }
+}
+
+#[derive(Clone, Debug)]
+pub struct MockServer {
+  mock_execution: MockExecution,
+  next_operation_response: Arc<Mutex<usize>>,
+  received_messages: Arc<Mutex<Vec<(String, Box<protobuf::Message>)>>>,
+}
+
+impl MockServer {
+  fn new(mock_execution: MockExecution) -> MockServer {
+    MockServer {
+      mock_execution: mock_execution,
+      next_operation_response: Arc::new(Mutex::new(0)),
+      received_messages: Arc::new(Mutex::new(vec![])),
+    }
+  }
+
+  fn log<T: protobuf::Message + Sized>(&self, message: T) {
+    self.received_messages.lock().unwrap().push((message.descriptor().name().to_string(), Box::new(message)));
+  }
+
+  fn display_all<T: Debug>(iter: Iter<T>) -> String {
+    iter.map(|i| format!("{:?}\n", i)).collect::<Vec<_>>().concat()
+  }
+
+  fn send_next_operation(&self, sink: grpcio::UnarySink<super::bazel_protos::operations::Operation>) {
+    let mut next_operation_response = self.next_operation_response.deref().lock().unwrap();
```

Comment:
Would it be terribly awkward to just make this an `Arc<Mutex<VecDeque<Operation>>>`, and just consume/drop them as you go? I love immutability in general, but given that we can't have it here, using a queue is potentially clearer.

Reply:
Yeah, I started off with that, but decided immutability (and being able to inspect what was expected without having to copy the Vec) seemed nice. I'll switch back :)